util/esp32_httpd_vfs.c

```diff
@@ -0,0 +1,258 @@
+#include "esp_log.h"
+#include "libesphttpd/esp.h"
+#include "libesphttpd/httpd.h"
+
+// If the client does not advertise that he accepts GZIP send following warning message (telnet users for e.g.)
+static const char *gzipNonSupportedMessage = "HTTP/1.0 501 Not implemented\r\nServer: esp8266-httpd/"HTTPDVER"\r\nConnection: close\r\nContent-Type: text/plain\r\nContent-Length: 52\r\n\r\nYour browser does not accept gzip-compressed data.\r\n";
+
+
+//This is a catch-all cgi function. It takes the url passed to it, looks up the corresponding
+//path in the filesystem and if it exists, passes the file through. This simulates what a normal
+//webserver would do with static files.
+CgiStatus ICACHE_FLASH_ATTR cgiEspVfsHook(HttpdConnData *connData) {
+	FILE *file=connData->cgiData;
+	int len;
+	char buff[1024];
+	char filename[256];
+	char acceptEncodingBuffer[64];
+	int isGzip;
+	bool isIndex = false;
+	struct stat filestat;	
+
+	if (connData->isConnectionClosed) {
+		//Connection aborted. Clean up.
+		fclose(file);
+		return HTTPD_CGI_DONE;
+	}
+
+	//First call to this cgi.
+	if (file==NULL) {
+		if (connData->cgiArg != NULL) {
+			strcpy(filename, connData->cgiArg);
+		} 
+		strcat(filename, connData->url);
+		ESP_LOGI(__func__, "GET: %s", filename);
+		
+		if(filename[strlen(filename)-1]=='/') filename[strlen(filename)-1]='\0';
+		if(stat(filename, &filestat) == 0) {
+			if((isIndex = S_ISDIR(filestat.st_mode))) {
+				strcat(filename, "/index.html");
+			}
+		}
+		
+		file = fopen(filename, "r");
+		isGzip = 0;
+		
+		if (file==NULL) {
+			// Check if requested file is available GZIP compressed ie. with file extension .gz
+		
+			strcat(filename, ".gz");
+			ESP_LOGI(__func__, "GET: GZIPped file - %s", filename);
+			file = fopen(filename, "r");
+			isGzip = 1;
+			
+			if (file==NULL) {
+				return HTTPD_CGI_NOTFOUND;
+			}				
+		
+			// Check the browser's "Accept-Encoding" header. If the client does not
+			// advertise that he accepts GZIP send a warning message (telnet users for e.g.)
+			httpdGetHeader(connData, "Accept-Encoding", acceptEncodingBuffer, 64);
+			if (strstr(acceptEncodingBuffer, "gzip") == NULL) {
+				//No Accept-Encoding: gzip header present
+				httpdSend(connData, gzipNonSupportedMessage, -1);
+				fclose(file);
+				return HTTPD_CGI_DONE;
+			}
+		}
+
+		connData->cgiData=file;
+		httpdStartResponse(connData, 200);
+		httpdHeader(connData, "Content-Type", isIndex?httpdGetMimetype("index.html"):httpdGetMimetype(connData->url));
+		if (isGzip) {
+			httpdHeader(connData, "Content-Encoding", "gzip");
+		}
+		httpdHeader(connData, "Cache-Control", "max-age=3600, must-revalidate");
+		httpdEndHeaders(connData);
+		return HTTPD_CGI_MORE;
+	}
+
+	len=fread(buff, 1, 1024, file);
+	if (len>0) httpdSend(connData, buff, len);
+	if (len!=1024) {
+		//We're done.
+		fclose(file);
+		return HTTPD_CGI_DONE;
+	} else {
+		//Ok, till next time.
+		return HTTPD_CGI_MORE;
+	}
+}
+
+
+//cgiEspVfsTemplate can be used as a template.
```

Comment:
could you elaborate the intent here? is the user supposed to customize this function? when would they want to do so?

Reply:
Delete